### PR TITLE
Throw error explicitly if atom is not provided.

### PR DESCRIPTION
**Pull request by alexandrzavalii**
When you call the hook without an atom/selector. `useRecoilState()`
It will throw the error `Error: Cannot read property 'key' of undefined`

Maybe we could check if the atom or selector is provided, if not then throw a meaningful error.
And the check could be applied to all other hooks as well.

Please let me know what do you think. :)

### CODE REVIEW

File: src/hooks/Recoil_Hooks.js [hunk 1]
```diff
@@ -232,6 +233,7 @@ function useInterface(): RecoilInterface {
     function useRecoilState<T>(
       recoilState: RecoilState<T>,
     ): [T, SetterOrUpdater<T>] {
+      if (!isRecoilValue(recoilState)) throw new Error('Invalid atom provided');
```
Comment (drarmstr) [inline src/hooks/Recoil_Hooks.js:236]: nodes may be either atoms or selectors.  the type is `RecoilState<T>`

**Comment by drarmstr** [comment]
We should handle all hooks consistently.

cc @habond - What we were discussing..

File: src/hooks/Recoil_Hooks.js (hunk 1 above)
Comment (alexandrzavalii) [reply src/hooks/Recoil_Hooks.js]: yes, isRecoilValue should be checking for RecoilState.
Maybe the error message should be? `No atom or selector provided`

**Comment by alexandrzavalii** [comment]
@drarmstr yes we should add the similar check to all other hooks.

File: src/hooks/Recoil_Hooks.js [hunk 2]
```diff
@@ -232,6 +233,8 @@ function useInterface(): RecoilInterface {
     function useRecoilState<T>(
       recoilState: RecoilState<T>,
     ): [T, SetterOrUpdater<T>] {
+      if (!isRecoilValue(recoilState))
+        throw new Error('No atom or selector provided');
```
Comment (acutmore) [inline src/hooks/Recoil_Hooks.js:237]: Perhaps these checks could be wrapped in a check for `__DEV__` to keep the bundle size down in production.

Comment (alexandrzavalii) [reply src/hooks/Recoil_Hooks.js:237]: @acutmore good idea!

**Comment by alexandrzavalii** [comment]
@drarmstr @acutmore Thanks for taking a look at it! :) 
Let me know if the check looks okay, and I will add it to all the hooks.

File: src/hooks/Recoil_Hooks.js [hunk 3]
```diff
@@ -232,6 +233,11 @@ function useInterface(): RecoilInterface {
     function useRecoilState<T>(
       recoilState: RecoilState<T>,
     ): [T, SetterOrUpdater<T>] {
+      if (__DEV__) {
+        if (!isRecoilValue(recoilState))
```
Comment (drarmstr) [inline src/hooks/Recoil_Hooks.js:237]: Looks like we don't have prettier running automatically on the GitHub side.  All if statements need blocks, single-statements aren't allowed.

File: src/hooks/Recoil_Hooks.js [hunk 4]
```diff
@@ -232,6 +233,11 @@ function useInterface(): RecoilInterface {
...
+          throw new Error('No atom or selector provided');
```
Comment (drarmstr) [inline src/hooks/Recoil_Hooks.js:238]: Hmm... maybe not clear that what was passed was invalid vs needing something else to be passed in - maybe "Invalid parameter, atom or selector expected"

File: src/hooks/Recoil_Hooks.js [hunk 5]
```diff
@@ -208,12 +218,18 @@ function useInterface(): RecoilInterface {
...
+      }
       return () => setRecoilValue(storeRef.current, recoilState, DEFAULT_VALUE);
     }
 
     function useRecoilValueLoadable<T>(
       recoilValue: RecoilValue<T>,
     ): Loadable<T> {
+      if (__DEV__) {
+        validateRecoilState(recoilState);
+      }
```
Comment (alexandrzavalii) [inline src/hooks/Recoil_Hooks.js:232]: `useRecoilValueLoadable` is used by `useRecoilValue` which is used by `useRecoilState`

**Comment by drarmstr** [comment]
Local edits on Phabricator:
* Name `validateRecoilValue()` for type consistency
* Fix `useRecoilValueLoadable` to pass `recoilValue` instead of `recoilState`
* Remove references to `console`
* Update test to use custom component that catches error to check if contents have the word parameter to see if it has a meaningful message since an error will be thrown either way.
* Misc lint

**Comment by davidmccabe** [comment]
This is a good change. I would prefer if the message was: `Invalid argument to ${hookName}: expected an atom or selector but got ${serializedParameter}.` as this explains what's going on a bit more thoroughly and would make debugging the problem a little faster. Thanks so much for contributing this!

**Comment by drarmstr** [comment]
Updated error message on Phabricator, no need to update PR.

**Pull request closed without merging**